docs/advanced-topics/custom-vertex-id.md

```diff
@@ -0,0 +1,89 @@
+
+Simply set `graph.set-vertex-id=<true/false>` and `graph.allow-string-vid=<true/false>` in your config file.
+When the graph is created, this value is loaded and persisted. After the graph is created,
+changing the option value in the config file would make no difference as JanusGraph simply ignores the value.
+
+### Alter an existing Graph
+
+To alter a global offline config (not limited to the aforementioned two options), you need to shut
+down all JanusGraph instances. Then open a gremlin console, connect to JanusGraph instance and run
+the following:
+
+```groovy
+mgmt = graph.openManagement();
+mgmt.set("graph.set-vertex-id", true);
+// optional, if you want to provide string ID
+mgmt.set("graph.allow-string-vid", true);
+mgmt.commit();
+```
+
+It will return an error if you have open JanusGraph instances. Otherwise, it will succeed and
+now your global configs would have been updated.
+
+## Custom Long ID
+
+To provide a long vertex ID, you must call `graph.getIDManager().toVertexId(long)` to
+retrieve a transformed JanusGraph ID. The input number must be positive. This is because
+certain range of long IDs are reserved for internal usage, so you have to always call
+the aforementioned utility to convert your ID. Unfortunately, this is not supported in
+all gremlin clients as it is a JanusGraph specific utility. You can create vertices like
+below:
+
+```groovy
+g = graph.traversal()
+g.addV().property(T.id, graph.getIDManager().toVertexID(123L)).next()
+g.tx().commit()
+```
+
+Note that if you want to get back the original ID you provided, you need to call
+`graph.getIDManager().fromVertexID(long)`.
+
+## Custom String ID
+
+A string vertex ID can have arbitrary length as long as the underlying storage backend
+permits. The string value must consist of ASCII characters only, and it cannot contain
+JanusGraph reserved relation delimiter. By default, this reserved character is `-` (dash),
+which means you cannot use UUID string as a vertex ID since it contains a `-` character.
+However, JanusGraph allows you to set `JANUSGRAPH_RELATION_DELIMITER` system property
+which can be any visible ASCII character. Once it is set, JanusGraph will prohibit that
+specific character instead of the default `-` character. For example, you can use the
+following command to set system property when opening gremlin console:
+
+```bash
+JAVA_OPTIONS="-DJANUSGRAPH_RELATION_DELIMITER=@" ./bin/gremlin.sh
+```
+
+and then you should be able to use UUID as custom id since now the reserved character
+becomes `@` rather than `-`.
+
+```groovy
+g = graph.traversal()
+g.addV().property(T.id, UUID.randomUUID().toString()).next()
+g.addV().property(T.id, "custom_vid_001").next()
+g.tx().commit()
+```
+
+!!! warning
+    If you decide to replace default `JANUSGRAPH_RELATION_DELIMITER`, you must
+    always set the system property value consistently across all JanusGraph
+    instances, including servers and clients. Otherwise, data corruption could
+    happen.
```

Comment:
Thank you @li-boxuan for the great work!
I didn't start the review yet, but wanted to leave thoughts / questions below.
1. I didn't research it myself, but as you worked in this direction already, I wonder if adding possibility to use raw bytes as vertex ids would make sense or possible at all.  I know that many databases support using raw bytes as primary keys (for example, `blob` in Cassandra / ScyllaDB or  `Binary` in HBase). I suppose using raw bytes could have more benefit then using String as users might literally serialize anything into such ids (UUIDs, integers, custom strings, custom objects, etc.). I suppose, having stored ids as bytes eliminates the need of using specific string characters and also gives ability to serialize more efficiently.
2. I just wander what is the reason to have it as a system property and not JanusGraph `GLOBAL_OFFLINE` configuration property?

Reply:
> I wonder if adding possibility to use raw bytes as vertex ids would make sense or possible at all

Good question. Unfortunately, it's not possible because of the data layout in JanusGraph. 

![image](https://user-images.githubusercontent.com/25746010/232663880-cd3b8bb7-42e6-4efd-900e-24e26b2ebae3.png)

JanusGraph needs a way to mark the boundary of different components. For example, it needs to know the end of "adjacent vertex id" and the start of "edge id". When using a long-type ID (which is the status quo), JanusGraph uses a special bit as a marker. Specifically, any byte with the first bit being 1 is a boundary marker.

When using a string-type ID (which is in this PR), we use a special byte to mark the boundary. Since we require the string ID to consist of ASCII characters, we are able to easily use a special byte to mark the boundary.

What about byte IDs? It's impossible unless we require users to avoid using a specific byte.

Moreover, IIRC, Elasticsearch requires document ID to be of string type. We could use base64 to encode the user-given ID but it will be a huge breaking change in `RelationalIdentifier`. Your second question is related.

> what is the reason to have it as a system property and not JanusGraph GLOBAL_OFFLINE configuration property

Unfortunately, the delimiter is not only used by JanusGraph server code but also by driver code. In the driver code, you don't have access to any JanusGraph configuration. Using system property to control is definitely not ideal but I am afraid there is nothing else we could do unless we introduce a breaking change (e.g. always use base64 to encode the relation identifier), which will make JanusGraph NOT backward-compatible :(

Reply:
Thank you for the great explanation! Definitely makes sense. 
I assume that the default delimiter `-` is chosen to not have breaking changes in relationship ids encoding, but in case I'm wrong and it's not a breaking change then I guess some less popular character would make more sense to be used as a default one (like `'` or `*`). 
I will try to check the PR later this week.

Reply:
> I assume that the default delimiter - is chosen to not have breaking changes in relationship ids encoding

You are right - we have to keep it for backward compatibility.

Reply:
> What about byte IDs? It's impossible unless we require users to avoid using a specific byte.

I guess if we encode byte array size (4 bytes at the beginning of adjacent vertex id byte array) then it could be possible to read those 4 bytes and understand when the specific key is finishing. This would, of course, increase the key size +3 bytes. It just some thoughts for future in case we ever want to implement that vertex ids as `byte[]`.

Reply:
> I guess if we encode byte array size (4 bytes at the beginning of adjacent vertex id byte array) then it could be possible to read those 4 bytes and understand when the specific key is finishing

That makes sense, yes, I think it should work. And yes, we could make it happen in the future. Currently, we leverage the fact that a long-type ID always have the first bit of the first byte being 0. I use an arbitrary byte that starts with bit 1 to mark that it's a string-type ID. In the future we could use another byte to mark that it's a different type.